### PR TITLE
Lenngren-based X25519 for non-alt ARM code

### PR DESCRIPTION
**Pull request by jargh**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by aqjune-aws** [comment]
Hi, since the updated functions now use Neon, should we update test/benchmark.c to run them only when neon is enabled?

Or should we simply assume that Arm always has Neon? I am not sure whether AWS-LC can rely on this assumption, though.

**Comment by Emill** [comment]
Reference: https://developer.arm.com/documentation/102525/0100/Compiling-for-Neon-with-Arm-Compiler-6:

> Neon is required in all standard Armv8-A implementations, so targeting any Armv8-A architecture or processor will allow the generation of Neon code.

Not sure about Armv9-A though.

**Comment by aqjune-aws** [comment]
I checked these things:
- The updates to the main specifications `CURVE25519_X25519_SUBROUTINE_CORRECT` and `CURVE25519_X25519_BYTE_SUBROUTINE_CORRECT` contained minor changes such as used stack size/code size/modifiable register set, and I thought these were okay
- The update to the precise definition of callee-save registers regarding Q8~Q15 (thanks.)
- The theorems printed in the proofs CI are okay (did not change compared to the master branch)
- The update to the postcondition of `BIGNUM_INV_P25519_SUBROUTINE_CORRECT` which looks correct to me.
```
(\s. read PC s = returnaddress /\
                  (coprime(p_25519,n)
                   ==> bignum_from_memory(z,4) s = inverse_mod p_25519 n))
```
=>
```
(\s. read PC s = returnaddress /\
                  bignum_from_memory(z,4) s =
                  (if p_25519 divides n then 0 else inverse_mod p_25519 n))
```

I will create a follow-up pull request that removes the existing Neon check condition from test.c/benchmark.c .

I think the lemmas `SIMD_SPLIT_JOIN_CLAUSES`, `EXTRA_SPLIT_JOIN_CLAUSES` can probably be used in other places using SIMD registers in s2n-bignum as well.

**Review by aqjune-aws — APPROVED** [review]

**Comment by aqjune-aws** [comment]
If this is merged, AWS-LC might want to link these assembly files only when Neon is supported. In theory this might be not necessary, but wanted to let people know :)

File: arm/curve25519/curve25519_x25519.S
Comment (hanno-becker) [inline arm/curve25519/curve25519_x25519.S:10]: Can we change this to "Abdulrahman, Becker, Kannwischer, Klein"?

File: arm/curve25519/curve25519_x25519.S
Comment (hanno-becker) [inline arm/curve25519/curve25519_x25519.S:205]: Can we say "... starting from the un-interleaved developed by Becker et. al in the SLOTHY paper", or just "... developed in the SLOTHY paper [ref]"?

**Comment by hanno-becker** [comment]
This is amazing, thank you @jargh.

File: arm/curve25519/curve25519_x25519.S
Comment (jargh) [reply arm/curve25519/curve25519_x25519.S:205]: Thank you June for the additional followup work and the careful review!

I will fix those attributions as Hanno suggests now before we merge.

**Comment by jargh** [comment]
I've tweaked the attributions, which will trigger a re-run of all checks.
Assuming nothing untoward happens, I'll merge after that.

**Pull request merged**
